src/common/spawn-primary-client.c

```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Copied from https://github.com/cage-kiosk/
+ *
+ * Copyright (c) 2018-2020 Jente Hidskes
+ * Copyright (c) 2019 The Sway authors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#define _POSIX_C_SOURCE 200809L
+#include <assert.h>
+#include <fcntl.h>
+#include <glib.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+#include <sys/wait.h>
+#include <unistd.h>
+#include <wlr/util/log.h>
+#include "common/spawn.h"
+#include "common/spawn-primary-client.h"
+#include "labwc.h"
+
+static int
+sigchld_handler(int fd, uint32_t mask, void *data)
```

Comment:
Could we call that differntly? This is not a `SIGCHLD` handler but rather a usual fd watcher that watches for `WL_EVENT_HANGUP | WL_EVENT_ERROR`.

So maybe something like `handle_pipe_error()` instead.

I am also not convinced that we should use the exact cage implementation and not just write our own in `common/spawn.c`. IMHO it basically only clutters the src/common/ directory and is almost the same code that we already have in `spawn.c`. Maybe we could simply change the `spawn_async_no_shell()` function to accept an optional `int *err_pipe` argument and then create a small wrapper function that calls `spawn_sync_no_shell()` and sets up the error handler?